### DISCUSSION TITLE
Example of Enriching Logs with Serilog?

### DISCUSSION THREAD

**Discussion by JohnGalt1717** [microsoft/aspire#1803] [General]
Is there any guidance on this? How should this be implemented?

**Comment by davidfowl** [comment]
What's aspire specific about serilog enrichment?

**Comment by JohnGalt1717** [comment]
The Host, and then how to make all of your services use shared configuration for serilog since the extension method doesn't have access to the IHost.

**Comment by davidfowl** [comment]
Ah I see, you're asking about the extension methods that only support IHostApplicationBuidler and not IHostBuilder. For serilog, you might need to use the other method of wiring it up. https://github.com/serilog/serilog-extensions-logging

cc @eerhardt

**Comment by eerhardt** [comment]
Can you call `builder.Logging.AddSerilog(...);`?

**Comment by JohnGalt1717** [comment]
you need more than that. You have to create the root log something like this:

```
		Log.Logger = new LoggerConfiguration()
			.MinimumLevel.Override("Microsoft", LogEventLevel.Information)
			.Enrich.FromLogContext()
			.WriteTo.Console()
			.CreateBootstrapLogger();

```

Before that call won't error.

Which has to happen for every microservice.

Then you have to add something like this:

```
		builder.Services.AddSerilog((IServiceProvider sp, LoggerConfiguration configuration) => {
			configuration.ReadFrom.Configuration(builder.Configuration);
			configuration.ReadFrom.Services(sp);
			configuration.Enrich.FromLogContext();
			configuration.WriteTo.Console();
		});

```

But I still can't get the host to just output serilog messages even when I set that up on the host program.cs. It works fine for the microservices though. (and I just created a shared static class that creates the slimbuilder and does the log stuff at the beginning and then put the AddSerialog in the extension method.

Was hoping there was a more elegant way, or something I was missing.

**Comment by eerhardt** [comment]
> But I still can't get the host to just output serilog messages even when I set that up on the host program.cs. 

You mean the AppHost.csproj's Program.cs? Is there a reason you want to "enrich" this log? This project isn't technically part of your app - it doesn't get deployed to the cloud. It is more of an "orchestrator" project to tell the tooling which microservice projects and containers to spin up.

**Comment by JohnGalt1717** [comment]
It generates noise like crazy that is scrolling by endlessly and since I use serilog everywhere else I was hoping to use it to silence what doesn't seem to silence by setting the loglevel in the appsettings.config

**Comment by davidfowl** [comment]
I'm confused about this too. Can you help us understand why serilog on the apphost project is important?

**Comment by JohnGalt1717** [comment]
The host when it runs opens a new terminal and spews stuff like this:

```
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint 'gRPC - /opentelemetry.proto.collector.metrics.v1.MetricsService/Export'
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'gRPC - /opentelemetry.proto.collector.metrics.v1.MetricsService/Export'
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/2 POST http://localhost:7064/opentelemetry.proto.collector.metrics.v1.MetricsService/Export - 200 - application/grpc 5.3012ms
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/2 POST http://localhost:7064/opentelemetry.proto.collector.metrics.v1.MetricsService/Export - application/grpc -
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/2 POST http://localhost:7064/opentelemetry.proto.collector.metrics.v1.MetricsService/Export - application/grpc -
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint 'gRPC - /opentelemetry.proto.collector.metrics.v1.MetricsService/Export'
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint 'gRPC - /opentelemetry.proto.collector.metrics.v1.MetricsService/Export'
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'gRPC - /opentelemetry.proto.collector.metrics.v1.MetricsService/Export'
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'gRPC - /opentelemetry.proto.collector.metrics.v1.MetricsService/Export'
```

Pretty much continuously.

In my configuration I have this:
```

	"Logging": {
		"LogLevel": {
			"Microsoft.AspNetCore.Hosting.Diagnostics": "Warning"
		}
	},
```

Which is being ignored.

And the host for Aspire doesn't have access to the Logging stuff that the rest of the IHost extensions have.

**Comment by DylanPMunyard** [comment]
@JohnGalt1717 inside `AppHost\Program.cs` this works for me:

```csharp
builder.Services.AddSerilog(configuration =>
{
    configuration.MinimumLevel.Override("Aspire.Hosting.Dcp", LogEventLevel.Warning);
    configuration.WriteTo.Console();
});
```

Are you referring to the `UseSerilog` extension not available on `IDistributedApplicationBuilder`? It just calls `AddSerilog` under the covers so you could write your own extension?

The bootstrap logger works for me too:

```csharp
Log.Logger = new LoggerConfiguration()
    .WriteTo.Console()
    .CreateBootstrapLogger();

builder.Services.AddSerilog(configuration =>
{
    configuration.MinimumLevel.Override("Aspire.Hosting.Dcp", LogEventLevel.Warning);
    configuration.WriteTo.Console();
});

Log.Information("Aspire.Host has started")
```

```
[11:48:15 INF] Aspire.Host has started
[11:48:15 INF] Now listening on: http://localhost:15015
[11:48:15 INF] OTLP server running at: http://localhost:16130
[11:48:17 INF] Application started. Press Ctrl+C to shut down.
[11:48:17 INF] Hosting environment: Development
[11:48:17 INF] Content root path: C:\Users\dmuny\projects\home-automation-aspire\home-automation-aspire.AppHost
```

**Comment by davidfowl** [comment]
I guess i don't see the purpose of using serilog in the host. Why does this matter? Or is this just a question about how one would ago about doing that just generally?

**Reply by DylanPMunyard** [reply]
In my case I was plumbing all my logs to sink to Seq, including the app host. So it wasn't so much a case of enriching as where to send the logs to.

**Comment by BartNetJS** [comment]
I'm struggling to get SeriLog and Aspire structured logs work together.
I also tried the above suggestions.
I've added a question at Serilog github, but maybe this is better place.
See https://github.com/serilog/serilog-aspnetcore/issues/359

**Reply by gmij** [reply]
I feel that the approach of this post is heavily skewed towards using OTLP, whereas Aspire's direct integration with Seq appears to be more optimal. Wouldn't it be more effective if Serilog could directly replace the default logger, rather than routing through OTLP with Serilog? Use Serilog WriteTo OTLP approach raises an issue of numerous log submission requests in Aspire or Seq's logs.

**Comment by alex-todorov-j2** [comment]
The https://github.com/serilog/serilog-aspnetcore/issues/359 approach works. 

But the reason to pick Serilog is that the ILogger<T> implementation in Microsoft.Extensions.Logging doesn't really give you structured logging. You have to serialize complex types first.  

In .NET Aspire, ServiceDefaults project, Extensions  class, method ConfigureOpenTelemetry, there is this call:

```
builder.Logging.AddOpenTelemetry(logging =>
{
    logging.IncludeFormattedMessage = true;
    logging.IncludeScopes = true;
});   
```
Can this be configured for structured logging? So that you don't have to add Serilog or another external provider.
